runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -0,0 +1,443 @@
+ */
+public class FlinkStreamingPortablePipelineTranslator implements FlinkPortablePipelineTranslator<
+        FlinkStreamingPortablePipelineTranslator.StreamingTranslationContext> {
+
+  /**
+   * Creates a streaming translation context. The resulting Flink execution dag will live in a new
+   * {@link StreamExecutionEnvironment}.
+   */
+  public static StreamingTranslationContext createTranslationContext(JobInfo jobInfo) {
+    PipelineOptions pipelineOptions;
+    try {
+      pipelineOptions = PipelineOptionsTranslation.fromProto(jobInfo.pipelineOptions());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    StreamExecutionEnvironment executionEnvironment =
+            FlinkExecutionEnvironments.createStreamExecutionEnvironment(
+                    pipelineOptions.as(FlinkPipelineOptions.class));
+    return new StreamingTranslationContext(jobInfo, pipelineOptions, executionEnvironment);
+  }
+
+  /**
+   * Streaming translation context. Stores metadata about known PCollections/DataStreams and holds
+   * the Flink {@link StreamExecutionEnvironment} that the execution plan will be applied to.
+   */
+
+  public static class StreamingTranslationContext
+          implements FlinkPortablePipelineTranslator.TranslationContext {
+
+    private final JobInfo jobInfo;
+    private final PipelineOptions options;
+    private final StreamExecutionEnvironment executionEnvironment;
+    private final Map<String, DataStream<?>> dataStreams;
+
+    private StreamingTranslationContext(
+            JobInfo jobInfo,
+            PipelineOptions options,
+            StreamExecutionEnvironment executionEnvironment) {
+      this.jobInfo = jobInfo;
+      this.options = options;
+      this.executionEnvironment = executionEnvironment;
+      dataStreams = new HashMap<>();
+
+    }
+
+    @Override
+    public JobInfo getJobInfo() {
+      return jobInfo;
+    }
+
+    public PipelineOptions getPipelineOptions() {
+      return options;
+    }
+
+    public StreamExecutionEnvironment getExecutionEnvironment() {
+      return executionEnvironment;
+    }
+
+    public <T> void addDataStream(String pCollectionId, DataStream<T> dataSet) {
+      dataStreams.put(pCollectionId, dataSet);
+    }
+
+    public <T> DataStream<T> getDataStreamOrThrow(String pCollectionId) {
+      DataStream<T> dataSet = (DataStream<T>) dataStreams.get(pCollectionId);
+      if (dataSet == null) {
+        throw new IllegalArgumentException(
+                String.format("Unknown datastream for id %s.", pCollectionId));
+      }
+      return dataSet;
+    }
+
+  }
+
+  interface PTransformTranslator<T> {
+    void translate(String id, RunnerApi.Pipeline pipeline, T t);
+  }
+
+  private final Map<String, PTransformTranslator<StreamingTranslationContext>>
+          urnToTransformTranslator = new HashMap<>();
+
+  FlinkStreamingPortablePipelineTranslator() {
+    urnToTransformTranslator.put(PTransformTranslation.FLATTEN_TRANSFORM_URN,
```

Comment:
Use ImmutableMap.of for thread safety reasons. HashMap has no concurrency guarantees.

Reply:
👍 we don't want to share translator instances between threads but it is cleaner anyways